src/main/java/org/apache/ibatis/type/ArrayTypeHandler.java

```diff
@@ -30,9 +31,25 @@ public ArrayTypeHandler() {
     super();
   }
 
+  @Override
+  protected void setNullParameter(PreparedStatement ps, int i, JdbcType jdbcType) throws SQLException {
+    ps.setNull(i, Types.ARRAY);
+  }
+  
   @Override
   public void setNonNullParameter(PreparedStatement ps, int i, Object parameter, JdbcType jdbcType) throws SQLException {
-    ps.setArray(i, (Array) parameter);
+    Array array = null;
+    if (parameter instanceof Array) {
+      array = (Array)parameter;
+    }
+    else {
+      if (!parameter.getClass().isArray()) {
+        throw new TypeException("ArrayType Handler requires SQL array or java array parameter and does not support type " + parameter.getClass());
+      }
+      Object[] values = (Object[])parameter;
+      array = ps.getConnection().createArrayOf(jdbcType.name(), values);
```

Comment:
OK I have modified the PR to free the array. 

IMHO the JDBC spec as well as the Array javadoc is confusing. They both talk about the Array being invalid after free(). In this case the Array *is* used further by the PreparedStatement, though  (the values of the Array haven't yet been sent to the database server). This is why I left out the call to free() at first.

Reply:
Thank you for the update and I realized that I should have been clearer (sorry!). 
If the parameter is an `Array`, we should let user code free it because it is possible that the same instance is referenced multiple times in the same query.

```java
if (parameter instanceof Array) {
  // no free() in this case
  ps.setArray(i, (Array)parameter);
} else {
  if (!parameter.getClass().isArray()) {
    throw new TypeException("ArrayType Handler requires SQL array or java array parameter and does not support type " + parameter.getClass());
  }
  String typeName = resolveTypeName(parameter.getClass().getComponentType());
  Array array = ps.getConnection().createArrayOf(typeName, (Object[])parameter);
  ps.setArray(i, array);
  array.free();
}
```
`resolveTypeName()` method is explained in the other comment.

Reply:
I have updated the branch with a type mapping in the form that you described. I have included the defaults for all regular java types but *not* for interface types like java.sql.Clob etc. Mapping those would require a different approach than a simple map and I think they'd be a very special case anyway.